app/app.go

```diff
@@ -926,6 +927,102 @@ func (app *App) FinalizeBlocker(ctx sdk.Context, req *abci.RequestFinalizeBlock)
 	return &resp, nil
 }
 
+func (app *App) DeliverTxWithResult(ctx sdk.Context, tx []byte) *abci.ExecTxResult {
+	deliverTxResp := app.DeliverTx(ctx, abci.RequestDeliverTx{
+		Tx: tx,
+	})
+	return &abci.ExecTxResult{
+		Code:      deliverTxResp.Code,
+		Data:      deliverTxResp.Data,
+		Log:       deliverTxResp.Log,
+		Info:      deliverTxResp.Info,
+		GasWanted: deliverTxResp.GasWanted,
+		GasUsed:   deliverTxResp.GasUsed,
+		Events:    deliverTxResp.Events,
+		Codespace: deliverTxResp.Codespace,
+	}
+}
+
+func (app *App) ProcessBlockSynchronous(ctx sdk.Context, txs [][]byte) []*abci.ExecTxResult {
+	txResults := []*abci.ExecTxResult{}
+	for _, tx := range txs {
+		txResults = append(txResults, app.DeliverTxWithResult(ctx, tx))
+	}
+	return txResults
+}
+
+// Returns a mapping of the accessOperation to the channels
+func getChannelsFromSignalMapping(signalMapping MessageCompletionSignalMapping) sdk.MessageAccessOpsChannelMapping {
+	channelsMapping := make(sdk.MessageAccessOpsChannelMapping)
+	for messageIndex, accessOperationsToSignal := range signalMapping {
+		for accessOperation, completionSignals := range accessOperationsToSignal {
+			var channels []chan interface{}
+			for _, completionSignal := range completionSignals {
+				channels = append(channels, completionSignal.Channel)
+			}
+			channelsMapping[messageIndex][&accessOperation] = channels
+		}
+	}
+	return channelsMapping
+}
+
+// TODO:: send signals to unblock and defer completion signal
+func (app *App) ProcessTxConcurrent(
+	ctx sdk.Context,
+	txIndex int,
+	txBytes []byte,
+	wg *sync.WaitGroup,
+	resultChan chan<- *abci.ExecTxResult,
+	txCompletionSignalingMap MessageCompletionSignalMapping,
+	txBlockingSignalsMap MessageCompletionSignalMapping,
+) {
+	defer wg.Done()
+	// Store the Channels in the Context Object for each transaction
+	ctx.WithTxBlockingChannels(getChannelsFromSignalMapping(txBlockingSignalsMap))
+	ctx.WithTxCompletionChannels(getChannelsFromSignalMapping(txCompletionSignalingMap))
+
+	// Deliver the transaction and store the result in the channel
+	resultChan <- app.DeliverTxWithResult(ctx, txBytes)
+}
+
+func (app *App) ProcessBlockConcurrent(
+	ctx sdk.Context,
+	txs [][]byte,
+	completionSignalingMap map[int]MessageCompletionSignalMapping,
+	blockingSignalsMap map[int]MessageCompletionSignalMapping,
+) []*abci.ExecTxResult {
+	var waitGroup sync.WaitGroup
+	resultChan := make(chan *abci.ExecTxResult)
+
+	// For each transaction, start goroutine and deliver TX
+	for txIndex, txBytes := range txs {
+		waitGroup.Add(1)
+		app.ProcessTxConcurrent(
+			ctx,
+			txIndex,
+			txBytes,
+			&waitGroup,
+			resultChan,
+			completionSignalingMap[txIndex],
+			blockingSignalsMap[txIndex],
+		)
+	}
+
+	// Waits for all the transactions to complete
+	go func() {
+		waitGroup.Wait()
```

Comment:
ah shoot, yeah good catch. We'll need another `waitGroup.Wait()` in the method too

Reply:
Ahh yeah, it's technically not really needed but more so it's just closing the channel. I guess in our case since we don't really need need to communicate that the data is done then we could just remove this